walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
+   */
+  private ArrayList<TimeRange> addTimeRange(ArrayList<TimeRange> occupiedTimeSlots,
+      TimeRange addedTimeSlot) {
+    for(TimeRange timeSlot : occupiedTimeSlots) {
+      // New TimeRange is already contained, so it is not added.
+      if(timeSlot.contains(addedTimeSlot)) {
+        return occupiedTimeSlots;
+      }
+      // New TimeRange contains an already existing meeting, so the new replaces the existing.
+      if(addedTimeSlot.contains(timeSlot)) {
+        occupiedTimeSlots.remove(timeSlot);
+        addTimeRange(occupiedTimeSlots, addedTimeSlot);
+        return occupiedTimeSlots;
+      }
+      // New TimeRange overlaps with an already existing meeting.
+      if(timeSlot.overlaps(addedTimeSlot)) {
+        addTimeRange(occupiedTimeSlots, mergeTimeRanges(timeSlot, addedTimeSlot));
+        return occupiedTimeSlots;
+      }
+    }
+    // No scenario applied to TimeRange added normally.
+    occupiedTimeSlots.add(addedTimeSlot);
+    return occupiedTimeSlots;
+  }
+
+  /*
+   * Merges overlapping TimeRanges into a single TimeRange that includes both of them.
+   */
+  private TimeRange mergeTimeRanges(TimeRange a, TimeRange b) {
+    return TimeRange.fromStartEnd(a.start() < b.start() ? a.start() : b.start(),
+        a.end() > b.end() ? a.end() : b.end(), false);
+  }
+
+  /*
+   * Returns a collection with TimeRanges that are not available for meetings.
+   */
+  private ArrayList<TimeRange> getOccupiedTimeSlots(Collection<Event> events,
+      HashSet<String> requestAttendees) {
+    ArrayList<TimeRange> timeRangeEventsRequestAttendees = new ArrayList<TimeRange>();
+    for(Event meeting : events) {
+      for(String attendee : meeting.getAttendees()) {
+        // If the meeting contains a requested attendee it is added.
+        if(requestAttendees.contains(attendee) != false) {
+          timeRangeEventsRequestAttendees = addTimeRange(timeRangeEventsRequestAttendees,
+              meeting.getWhen());
+          break;
+        }
+      }
+    }
+    Collections.sort(timeRangeEventsRequestAttendees, TimeRange.ORDER_BY_START);
+    return timeRangeEventsRequestAttendees;
+  }
+
+  /*
+   * Calculates whether requested meetings's duration fits in between occupied TimeRanges.
+   */
+  private Collection<TimeRange> getFreeTimeSlots(ArrayList<TimeRange> occupiedTimeSlots, 
+      long duration) {
+    Collection<TimeRange> possibleTimeSlots = new ArrayList<>();
+    // Return all available if there are no occupied TimeRanges.
+    if(occupiedTimeSlots.isEmpty()) {
+      possibleTimeSlots.add(TimeRange.WHOLE_DAY);
+      return possibleTimeSlots;
+    }
+    // From midnight to the first meeting.
+    if(occupiedTimeSlots.get(0).start() - TimeRange.START_OF_DAY >= duration) {
+      possibleTimeSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, occupiedTimeSlots.get(0).start(), false));
+    }
+    int previousEnd = occupiedTimeSlots.get(0).end();
+    int currentStart;
```

Comment:
Does this variable need to be declared outside the for loop? If it's never referenced outside the for loop, you can declare it as a local variable inside the loop!

Reply:
Ack!